llvm/include/llvm/ABI/Types.h

```diff
@@ -0,0 +1,543 @@
+      std::copy(VirtualBaseClasses.begin(), VirtualBaseClasses.end(),
+                VBaseArray);
+    }
+
+    ArrayRef<FieldInfo> FieldsRef(FieldArray, Fields.size());
+    ArrayRef<FieldInfo> BasesRef(BaseArray, BaseClasses.size());
+    ArrayRef<FieldInfo> VBasesRef(VBaseArray, VirtualBaseClasses.size());
+
+    return new (Allocator.Allocate<StructType>())
+        StructType(FieldsRef, BasesRef, VBasesRef, Size, Align, Pack, false,
+                   CXXRecord, Polymorphic, NonTrivialCopy, NonTrivialDtor,
+                   FlexibleArray, UnalignedFields, CanPassInRegister);
+  }
+  const StructType *
+  getCoercedStructType(ArrayRef<FieldInfo> Fields, TypeSize Size, Align Align,
+                       StructPacking Pack = StructPacking::Default,
+                       ArrayRef<FieldInfo> BaseClasses = {},
+                       ArrayRef<FieldInfo> VirtualBaseClasses = {},
+                       bool CXXRecord = false, bool Polymorphic = false,
+                       bool NonTrivialCopy = false, bool NonTrivialDtor = false,
+                       bool FlexibleArray = false, bool UnalignedFields = false,
+                       bool CanPassInRegister = false) {
+    FieldInfo *FieldArray = Allocator.Allocate<FieldInfo>(Fields.size());
+    std::copy(Fields.begin(), Fields.end(), FieldArray);
+
+    FieldInfo *BaseArray = nullptr;
+    if (!BaseClasses.empty()) {
+      BaseArray = Allocator.Allocate<FieldInfo>(BaseClasses.size());
+      std::copy(BaseClasses.begin(), BaseClasses.end(), BaseArray);
+    }
+
+    FieldInfo *VBaseArray = nullptr;
+    if (!VirtualBaseClasses.empty()) {
+      VBaseArray = Allocator.Allocate<FieldInfo>(VirtualBaseClasses.size());
+      std::copy(VirtualBaseClasses.begin(), VirtualBaseClasses.end(),
+                VBaseArray);
+    }
+
+    ArrayRef<FieldInfo> FieldsRef(FieldArray, Fields.size());
+    ArrayRef<FieldInfo> BasesRef(BaseArray, BaseClasses.size());
+    ArrayRef<FieldInfo> VBasesRef(VBaseArray, VirtualBaseClasses.size());
+    return new (Allocator.Allocate<StructType>())
+        StructType(FieldsRef, BasesRef, VBasesRef, Size, Align, Pack, false,
+                   CXXRecord, Polymorphic, NonTrivialCopy, NonTrivialDtor,
+                   FlexibleArray, UnalignedFields, CanPassInRegister, true);
+  }
+  const StructType *getUnionType(ArrayRef<FieldInfo> Fields, TypeSize Size,
+                                 Align Align,
+                                 StructPacking Pack = StructPacking::Default,
+                                 bool IsTransparent = false,
+                                 bool CanPassInRegs = false,
+                                 bool CXXRecord = false) {
+    FieldInfo *FieldArray = Allocator.Allocate<FieldInfo>(Fields.size());
+
+    for (size_t I = 0; I < Fields.size(); ++I) {
+      const FieldInfo &Field = Fields[I];
+      new (&FieldArray[I])
+          FieldInfo(Field.FieldType, 0, Field.IsBitField, Field.BitFieldWidth,
```

Comment:
Should we assert that the original fields all have an offset of 0 (so they can be directly copied)?

Reply:
Yeah, this is just intended as a sanity check. If we're trying to construct a union with non-zero offsets, something is wrong...